Make proxy shutdown when all connections are closed

save-buffer opened the pull request:
## Describe your changes
Adds functionality to the Proxy so that when it receives SIGTERM, it enters a state where it will automatically exit when the last connection to it closes. 
## Issue ticket number and link
#3333 
## Checklist before requesting a review
- [ x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

lassizci commented on proxy/src/main.rs:
```diff
@@ -116,11 +116,10 @@ async fn handle_signals() -> anyhow::Result<()> {
             _ = terminate.recv() => {
-                warn!("received SIGTERM, exiting immediately");
-                bail!("terminated");
+                warn!("received SIGTERM, shutting down once all existing connections have closed");
```
Because the current behavior is to shut down immediately on `SIGTERM` and it's also the default signal for `kill`, I think it would potentially cause less surprises if we kept the current behavior and made the graceful shutdown require something else (QUIT, INT).

Don't think there's really any global standard which one exactly as it's a bit different with for example nginx, haproxy, envoy, postgresql and so on. What do you think?

vadim2404 replied:
https://www.postgresql.org/docs/current/server-shutdown.html

It's a kind of inverted logic in my mind. Sigterm is used for "smart" shutdown.

funbringer replied:
I disagree with Lassi's proposal. I think we should stick with [what postgres does](https://www.postgresql.org/docs/current/server-shutdown.html).

vadim2404 commented on test_runner/regress/test_proxy.py:
```diff
@@ -105,3 +107,12 @@ def test_forward_params_to_client(static_proxy: NeonProxy):
+def test_close_on_connections_exit(static_proxy: NeonProxy):
+    with static_proxy.connect(options="project=irrelevant"):
+        with static_proxy.connect(options="project=irrelevant"):
+            if static_proxy._popen:
```
you can just use 

```python
assert static_proxy._popen
```

It will make `mypy` happy and add one more check in order to pass the test.

With this `if` statement, you will test nothing and `pytest` will mark this test as `passed`. IMO, it's a wrong behavior :)

save-buffer replied:
Good point, changed it

save-buffer commented on proxy/src/proxy.rs:
```diff
@@ -59,6 +61,23 @@ static NUM_BYTES_PROXIED_COUNTER: Lazy<IntCounterVec> = Lazy::new(|| {
+static EXIT_ON_ALL_CONNECTIONS_CLOSED: AtomicBool = AtomicBool::new(false);
+
+pub fn set_exit_on_connections_closed() {
+    EXIT_ON_ALL_CONNECTIONS_CLOSED.store(true, Ordering::Release);
```
Added a comment and switched to relaxed.

save-buffer commented on proxy/src/proxy.rs:
```diff
@@ -59,6 +61,23 @@ static NUM_BYTES_PROXIED_COUNTER: Lazy<IntCounterVec> = Lazy::new(|| {
+}
+
+pub fn exit_if_needed() {
+    if EXIT_ON_ALL_CONNECTIONS_CLOSED.load(Ordering::Acquire) {
```
Changed it to relaxed.

hlinnaka commented:
Do we still accept new connections after receiving SIGTERM? I guess Kubernetes will redirect new connections to a new instance, so in practice no new connections should arrive. Although I'm not sure what exactly the sequence is in Kubernetes, does it redirect new connections first, and only then send SIGTERM.

vadim2404 commented:
> Do we still accept new connections after receiving SIGTERM? I guess Kubernetes will redirect new connections to a new instance, so in practice no new connections should arrive. Although I'm not sure what exactly the sequence is in Kubernetes, does it redirect new connections first, and only then send SIGTERM.

it can possibly send new connections for some time to the old instance. Kubernetes doesn't guarantee that it stops to send new connection f

save-buffer commented on proxy/src/proxy.rs:
```diff
@@ -59,6 +61,23 @@ static NUM_BYTES_PROXIED_COUNTER: Lazy<IntCounterVec> = Lazy::new(|| {
+        let num_closed = NUM_CONNECTIONS_CLOSED_COUNTER.get();
+        if num_accepted == num_closed {
+            info!("All connections closed, terminating");
+            process::exit(0);
```
I decided cleanest way would just be to have a static CancellationToken in `proxy` which gets cancelled on `exit_if_needed`. It would be cleaner to have token be created in `proxy::task_main` and passed around, except having the `handle_signals` function also have a copy of the token is annoying.

save-buffer commented on test_runner/regress/test_proxy.py:
```diff
@@ -105,3 +107,12 @@ def test_forward_params_to_client(static_proxy: NeonProxy):
+
+def test_close_on_connections_exit(static_proxy: NeonProxy):
+    with static_proxy.connect(options="project=irrelevant"):
+        with static_proxy.connect(options="project=irrelevant"):
```
Removed them

save-buffer commented:
> it can possibly send new connections for some time to the old instance. Kubernetes doesn't guarantee that it stops sending new connections

Just added some logic to drop TCPListener so that we don't accept new connections after we receive signal.

cicdteam commented:
@save-buffer 
CC: @funbringer @hlinnaka @lassizci 

I think we need to take into account other features of using a proxy.

The proxy is behind the AWS Network Load Balancer (NLB), and NLB directly uses the IP addresses of the pods as targets. Kubernetes Services, Endpoints, kube-proxy - all these abstractions and components are NOT used in processing proxy traffic.

Some quotes from AWS Docs on how NLB ensures its targets work:

>Traffic is forwarded to the target group specified in the listener rule.

>You define health check settings for your load balancer on a per target group basis. Each target group uses the default health check settings, unless you override them when you create the target group or modify them later on. After you specify a target group in a rule for a listener, the load balancer continually monitors the health of all targets registered with the target group that are in an Availability Zone enabled for the load balancer. The load balancer routes requests to the registered targets that are healthy.

Let's see how health check works for 5432 proxy port. By default it uses TCP health checks (just try to make TCP connection to port). Health check interval - 10s, health check timeout - 10s, unhealthy threshold - 3 consecutive health check failures, healthy threshold - 3 consecutive health check successes.

By other words - NLB will continue pass new connections (and traffic of old connections) during 30s minimum after pods stop receiving traffic.

One more point - `Deregistration delay`:

>When you deregister a target, the load balancer stops creating new connections to the target. The load balancer uses connection draining to ensure that in-flight traffic completes on the existing connections. If the deregistered target stays healthy and an existing connection is not idle, the load balancer can continue to send traffic to the target. To ensure that existing connections are closed, you can do one of the following: enable the target group attribute for connection termination, ensure that the instance is unhealthy before you deregister it, or periodically close client connections.

That delay started when you deregister target (for example by AWS API call). In our case it happen when Pod goes in termination state (I hope). It managed by AWS ALB ingress controller.

>To change the amount of time that the load balancer waits before changing the state of a deregistering target to unused, update the deregistration delay value. We recommend that you specify a value of at least 120 seconds to ensure that requests are completed.

In our case it should be a week!  But default value of deregistration delay is 300s and range is 0-3600s.

Resuming - we need rethink our strategy about proxy shutdown. May be need implement some calls to AWS API to manage targets in some manner.

lassizci commented:
So I performed practical testing too to verify this works as planned, before we rolled that out initially. If we take a closer look what does the documentation say, I don't think it disagrees either:

> If the deregistered target stays healthy and an existing connection is not idle, the load balancer can continue to send traffic to the target. 

So this suggests that if these two conditions are met, NLB will keep sending traffic to the target. This was the result in my testing too:
1) The pod is in Terminating state, it is still running, has the ip address. NLB connects to the pod IP address and therefore the health check passes. 
2) I made sure something is sent over the connection once a second -> the connection qualifies as "not idle" (while not exactly sure what's the exact criteria)


>  To ensure that existing connections are closed, you can do one of the following: enable the target group attribute for connection termination, ensure that the instance is unhealthy before you deregister it, or periodically close client connections.

The point is that we _don't_ enable the "target group attribute for connection termination" and we _don't_ ensure the target is unhealthy before deregistering. So the deregister period expires (as there's a maximum) but the target remains healthy -> the target does disappear from the target list in AWS console, but NLB keeps sending traffic there as promised. The target won't be used for any new connections as it's no longer in the target list.

Once we get this in, we just won't have pods that _always_ hang for a week as they now do (probably even without any connections, as NLB should terminate the connection if it's idle).

lassizci commented:
> > it can possibly send new connections for some time to the old instance. Kubernetes doesn't guarantee that it stops sending new connections
> 
> Just added some logic to drop TCPListener so that we don't accept new connections after we receive signal.

Since we are running in Kubernetes, we should NOT stop accepting new connections once receiving the signal. Kubernetes stops sending new connections _eventually_ (pretty soon, but not soon enough). So if we rejected the connections, the clients would notice. We have long grace periods exactly to mitigate rejecting the client connections.

save-buffer commented:
> we should NOT stop accepting new connections once receiving the signal

Good point, I changed it back and added a part to the test to make sure that we do accept new connections.

koivunej commented on proxy/src/proxy.rs:
```diff
@@ -74,26 +101,33 @@ pub async fn task_main(
-            }),
-        );
+        }
     }
```
Proposing an alternative I've been thinking for pageserver.

It seems this acceptor task would be a good place to have a `tokio::task::JoinSet`, use that to spawn connection handlers. On cancellation, only stop the accepting, but then in a second loop just wait for all tasks to complete. Then when the main task has exited all connections have died.

No atomic counters are needed, but this `tokio::select` will need a separate listening on the joinset to cleanup any already handled connections. Internally it seemed like JoinSet uses a intrusive linked list, so it's implementation should be close to optimal aside from few bugs.

Didn't check that this task is ever awaited for. If it is already, then all should be good.

koivunej replied:
Pseudo-illustrative semi-suggestion:

```rust
pub async fn task_main(...) -> anyhow::Result<()> {

  let connections = tokio::task::JoinSet::new();

  loop {
    tokio::select! {
      accept_result = listener.accept() => {
        ...
        
        connections.spawn( // only changed
          async move {
            ...
          }
        );
      }
      _ = PROXY_CANCELLATION_TOKEN.cancelled() => {
        // stop accepting new connections and start waiting for old to die
        break;
      }
      maybe_exited = connections.join_next() => {
        // maintain the 
        match maybe_exited {
          Some(Ok(())) => { /* normal exit */ },
          Some(Err(e)) if e.is_cancelled() => unreachable!("we do not deschedule tasks which is different from them reacting to cancellation token and deciding to exit themselves"),
          Some(Err(e)) if e.is_panic() => { /* panic has already been logged by panic hook */ },
          Some(Err(e)) => { error!("unexpected error from joined connection: {e:?}); },
          None => { /* this happens when polled an empty connections, as in proxy is idle right now */ }
        }
      }
    }
  }
  
  // we've been asked to stop, let's just wait until connections complete
  while let Some(_) = connections.join_next().await {
    // pretty much the same match as above but I wrote it as an example,
    // can't see what should be done here (assuming no descheduling).
  }
  
  // all done
  Ok(())
}
```

koivunej replied:
This all could be juggled in a single tokio::select! with the preconditions, but I think it's simpler to have more verbose two loops.

koivunej replied:
Probably needs `drop(listener)` as well in the cancellation arm.

koivunej replied:
This has now been implemented, resolving..

funbringer commented:
@koivunej I see that you've already contributed to the discussion. May I ask you to continue reviewing this PR?

koivunej commented on proxy/src/proxy.rs:
```diff
@@ -124,8 +102,19 @@ pub async fn task_main(
+                    None => {
+                        if cancellation_token.is_cancelled()
+                        {
+                            drop(listener);
```
This is unnecessary (because automatic) if you are going to exit the fn on the next line.

save-buffer replied:
Removed

koivunej commented on proxy/src/proxy.rs:
```diff
@@ -124,8 +102,19 @@ pub async fn task_main(
+                        if cancellation_token.is_cancelled()
+                        {
+                            drop(listener);
+                            bail!("All connections closed, exiting");
```
I'd call this very much a normal exit assuming this is how the updates will done, no error needed, just `return Ok(())` or break out of the loop.

save-buffer replied:
I think we actually need an error here because we want Tokio to shut everything down when this loop exits

koivunej replied:
That is correct, didn't check the main. Resolving..

koivunej commented on proxy/src/proxy.rs:
```diff
@@ -124,8 +102,19 @@ pub async fn task_main(
+                    Some(Err(e)) => { bail!(e.to_string()) },
+                    None => {
+                        if cancellation_token.is_cancelled()
+                        {
```
Note that with-in `tokio::select` the `rustfmt` cannot help you, you need to manually do the formatting.

save-buffer replied:
Fixed

koivunej replied:
Woops, accidential unresolve. Resolved.

koivunej commented on proxy/src/proxy.rs:
```diff
@@ -124,8 +102,19 @@ pub async fn task_main(
+                match maybe_exited
+                {
+                    Some(Ok(())) => {},
+                    Some(Err(e)) => { bail!(e.to_string()) },
```
Also for future, `bail!(e.to_string())` is never a good idea because then the caller would only have the string to work with.

`anyhow::Error::new(e).context("what happened")` works as one alternative, or in here `return Err(e.into())` would work, but it's wrong for the fn itself.

save-buffer replied:
OK, I good point. I think I didn't fully understand your code at first. I just pasted it in

koivunej commented on proxy/src/main.rs:
```diff
@@ -66,11 +67,16 @@ async fn main() -> anyhow::Result<()> {
+        )),
         tokio::spawn(console::mgmt::task_main(mgmt_listener)),
     ];
 
```
New thread for `wss` handling (one line below this).

Do they need to keep proxy alive similar to `proxy::task_main`? If so, similar handling is needed, and we also probably want to group these "not infinite tasks" separatedly from the "infinite tasks" which we can just stop polling when we decide to exit.

save-buffer replied:
I rewrote the PR to implement the simple drain functionality we discussed offline, so marking this resolved. The new version does wait for web socket connections to terminate.

koivunej commented on proxy/src/main.rs:
```diff
@@ -66,39 +67,51 @@ async fn main() -> anyhow::Result<()> {
 
     // This combinator will block until either all tasks complete or
     // one of them finishes with an error (others will be cancelled).
     let tasks = tasks.into_iter().map(flatten_err);
-    let _: Vec<()> = futures::future::try_join_all(tasks).await?;
-
+    let (tasks_done, abort_handle) =
+        futures::future::abortable(futures::future::try_join_all(tasks));
+    let client_tasks_done =
+        futures::future::try_join_all(client_tasks.into_iter()).then(|_| async move {
+            abort_handle.abort();
+        });
+    let _ = tokio::join!(client_tasks_done, tasks_done);
     Ok(())
```
The `then` threw me off a bit.

In the previous design or this new one are we not actually cancelling any of what are now just `tasks`. What we end up aborting is the `try_join_all` of `Vec<tokio::task::JoinHandle>`, which is the same as doing nothing, but with an added unhandled `futures::stream::Aborted` `Err` value. Dropping a `tokio::task::JoinHandle` created by `tokio::spawn` **does nothing** (in the docs this is called "detaches the associated task").

So the statements in selection are equivalent to:

```suggestion
    futures::future::try_join_all(client_tasks.into_iter()).await
```

If we really wanted do de-schedule (or "abort") the `tasks`, then using `JoinSet` for the three would be the easiest way, because all unjoined tasks are by default de-scheduled on drop. I think this would be unnecessary for this PR which just adds graceful stop for the proxy connections. As in we don't need to stop responding via http for example, at least I am unaware of there being any state we should toggle on graceful stop.

Now I noticed the comment:

```rust
// This combinator will block until either all tasks complete or
// one of them finishes with an error (others will be cancelled).
```

I think my suggestion would achieve this because implicitly the rest of the `tokio::spawn`'ed tasks will be dropped when the `async fn main` returns due to dropping of the runtime.

I think there was some comment about a shutdown order being important as opposed your earlier version using `std::process::exit(0)` so I might need to update this comment.

koivunej replied:
> I think there was some comment about a shutdown order being important as opposed your earlier version using `std::process::exit(0)` so I might need to update this comment.

The comment was <https://github.com/neondatabase/neon/pull/3764#discussion_r1132493412>.

Noticed an error with my suggestion. I stand by that it is indeed a minimization of your latest changes, but I don't think it achieves what the original implementation achieved; now an error from any of the `tasks` would go unnoticed.

Easiest way would be to grow the "region of graceful stop" to the two tasks as well (mgmt handler, mgmt http api), then a single `try_join_all` (the original version) would work, as in "early exit on every `Err` but continue until all have completed `Ok(())`".

It might be worth-while to keep the signal handlers task out of this, to keep them just a separate `tokio::spawn`'ed task, because it's unsure if one really wants a signal handling task to ever complete with `Ok(())`.

koivunej replied:
Oki, I had misunderstood. We went over this on a call, and ended up with a simpler `tokio::select` race which handles:
- any of the "forever tasks" (in `tasks`) returning Err
- all of the "graceful shutdowning tasks (in `client_tasks`) returning Ok or Err

We couldn't use `tokio::try_join!` because the forever running tasks never fail, and also this way we don't need to have an "Ok" `Err` type in addition to `anyhow::Result<()>`.

koivunej reviewed (approved):
This is now looking good to me. We lack a test for wss graceful shutdown, but at least there's one for tcp.

Remember to remove the PR checklist from the squash merge commit.
